core/kernel/user_access.c

```diff
@@ -74,6 +96,137 @@ TEE_Result copy_to_user_private(void *uaddr, const void *kaddr, size_t len)
 	return res;
 }
 
+void *bb_alloc(size_t len)
+{
+	struct user_mode_ctx *uctx = get_current_uctx();
+	size_t offs = 0;
+	void *bb = NULL;
+
+	if (uctx && !ADD_OVERFLOW(uctx->bbuf_offs, len, &offs) &&
+	    offs <= uctx->bbuf_size) {
+		bb = uctx->bbuf + uctx->bbuf_offs;
+		uctx->bbuf_offs = ROUNDUP(offs, BB_ALIGNMENT);
+	}
+	return bb;
+}
+
+static void bb_free_helper(struct user_mode_ctx *uctx, vaddr_t bb, size_t len)
+{
+	vaddr_t bbuf = (vaddr_t)uctx->bbuf;
+
+	if (bb > bbuf && IS_ALIGNED(bb, BB_ALIGNMENT)) {
+		size_t prev_offs = bb - bbuf;
+
+		if (prev_offs + ROUNDUP(len, BB_ALIGNMENT) == uctx->bbuf_offs)
+			uctx->bbuf_offs = prev_offs;
+	}
+}
```

Comment:
I am curious if this is guaranteed to work

Reply:
This should work for freeing each buffer is done in reverse order. However, the main target is dealing with attributes in `core/tee/tee_svc_cryp.c` so it's might only the last allocation that should be freeable.